### PR TITLE
Airlock spawners 2: Electric Boogaloo

### PR DESCRIPTION
**Pull request by Citinited**
### Please note: this PR is a revival of #10942
---

**What does this PR do:**

This PR adds a new tool for mappers, the airlock spawner. What this does is that it will, upon `Initialize`, spawn its own airlocks, control equipment, and any vents that are required for the airlock to function. This means that you no longer need to create your own airlocks and tediously VV everything - it'll do it all for you.

The spawner will not create walls, floors, or a canister for you, it will leave that down to you. Simply connect your piping to the indicated pipe on the airlock and the rest will be done for you!

You can specify any size and orientation of the generated chamber and resultant airlocks. You could specify the entrance to be in the north and the exit to be in the east, for example.

This PR also adds some helpers, namely `NORTH_OF_TURF(turf/T)`, `EAST_OF_TURF(turf/T)`, `SOUTH_OF_TURF(turf/T)`, and `WEST_OF_TURF(turf/T)`. These seemed like they could be useful elsewhere so I stuck them in the appropriate file.

**Map changes**

None - these will come in a later PR, this is just to get them into the game first.

**Sprites**

How it looks in the map editor: 
image

1 by 1 airlock
image

1 by 2 airlock
image

2 by 2 airlock
image

3 by 2 airlock
image

3 by 3 airlock
image

Here is a video of it working fine and dandy
https://streamable.com/3g1gb

:cl:
add: Mappers have a new tool that creates a fully functional cycling airlock.
/:cl:

### CODE REVIEW

File: code/game/machinery/embedded_controller/airlock_controllers.dm [hunk 1]
```diff
@@ -76,6 +76,21 @@
 	name = "Airlock Controller"
 	tag_secure = 1
 
+/obj/machinery/embedded_controller/radio/airlock/airlock_controller/New(loc, given_id_tag, given_frequency, given_tag_exterior_door, given_tag_interior_door, given_tag_airpump, given_tag_chamber_sensor)
```
Comment (Citinited) [inline code/game/machinery/embedded_controller/airlock_controllers.dm]: Note to self: make this under `Initialize()`

**Comment by Markolie** [comment]
Merge posthaste!

File: code/ATMOSPHERICS/datum_pipeline.dm [hunk 2]
```diff
@@ -115,10 +117,12 @@ var/pipenetwarnings = 10
 
 /obj/machinery/atmospherics/proc/addMember(obj/machinery/atmospherics/A)
 	var/datum/pipeline/P = returnPipenet(A)
-	P.addMember(A, src)
+	if(P) //Sanity checking
+		P.addMember(A, src)
 
 /obj/machinery/atmospherics/pipe/addMember(obj/machinery/atmospherics/A)
-	parent.addMember(A, src)
+	if(parent && A) //Sanity checking
```
Comment (Fox-McCloud) [inline code/ATMOSPHERICS/datum_pipeline.dm]: I don't think these should ever be `null`.

This is one of those cases of "papering over the problem". 

Sure, it prevents the runtime, but it doesn't address the underlying problem

Comment (Citinited) [reply code/ATMOSPHERICS/datum_pipeline.dm]: This is fair enough. I'll see if I can get to the root of the cause

**Comment by AffectedArc07** [comment]
@variableundefined This PR doesnt actually change the maps, so it doesnt really need the `Map Edit` tag, since those are coming in a future PR
```
Map changes

None - these will come in a later PR, this is just to get them into the game first.
```

File: code/ATMOSPHERICS/datum_pipeline.dm (hunk 2 above)
Comment (Citinited) [reply code/ATMOSPHERICS/datum_pipeline.dm]: Okay I think this is fine if removed - I think it only runtimes if incorrectly mapped, anyway.

**Comment by MrMagolor** [comment]
Out of curiosity, are 3x3 airlocks used anywhere at all in the game currently? I don't think they are, but I could be wrong.

**Comment by Citinited** [comment]
The syndicate shuttle uses a 3 by 3 airlock - this is the only one that does, as far as I'm aware

**Comment by Citinited** [comment]
@Fox-McCloud @variableundefined I've fixed the runtime issue - turns out spawning multiple pipes without them calling `on_construction` first breaks things 😄 

I hope my code isn't *too* illegible

**Comment by Citinited** [comment]
🎉

**Pull request merged**